File: cpp/repl/basic2/evaluator.cpp
#include "evaluator.h"
#include "ast.h"

#include <iostream>

Evaluator::Evaluator() {

}

Object* Evaluator::eval(ASTNode *node, Environment *env) {
    switch(node->type) {
        case AST_PROGRAM: {
            return evalProgam(node, env);
            break;
        }
        case AST_EXPRESSION_STATEMENT: {
            return evalExpressionStatement(node, env);
            break;
        }
        case AST_INFIX_EXPRESSION: {
            return evalInfixExpression(node, env);
            break;
        }
        case AST_INTEGER: {
            return evalInteger(node, env);
            break;
        }
        default: {
            std::cout << "UNKNOWN EVALUATION OF AST NODE\n";
            break;
        }
    }
    return NULL;
}

Object* Evaluator::evalProgam(ASTNode *node, Environment *env) {
    ASTProgram *program = (ASTProgram*) node;
    for (int i = 0; i < program->statements.size(); i++) {
        Object *result = eval(program->statements.at(i), env);
        std::cout << result->inspect() << "\n";
    }
    return NULL;
}

Object* Evaluator::evalExpressionStatement(ASTNode *node, Environment *env) {
    ASTExpressionStatement *expressionStatement = (ASTExpressionStatement*) node;
    return eval(expressionStatement->expression, env);
}

Object* Evaluator::evalInfixExpression(ASTNode *node, Environment *env) {
    ASTInfixExpression *infixExpression = (ASTInfixExpression*) node;
    Object *left = eval(infixExpression->left, env);
    Object *right = eval(infixExpression->right, env);
    Object *result = NULL;
    if (left->type == INTEGER && right->type == INTEGER) {
        if (infixExpression->_operator == "+") {
            result = new IntegerObject();
            ((IntegerObject*) result)->value = ((IntegerObject*) left)->value + ((IntegerObject*) right)->value;
        }
        else if (infixExpression->_operator == "-") {
            result = new IntegerObject();
            ((IntegerObject*) result)->value = ((IntegerObject*) left)->value - ((IntegerObject*) right)->value;
        }
        else if (infixExpression->_operator == "*") {
            result = new IntegerObject();
            ((IntegerObject*) result)->value = ((IntegerObject*) left)->value * ((IntegerObject*) right)->value;
        }
        else if (infixExpression->_operator == "/") {
            result = new IntegerObject();
            ((IntegerObject*) result)->value = ((IntegerObject*) left)->value / ((IntegerObject*) right)->value;
        }
        else {
            std::cout << "UNKNOWN OPERATOR FOR INTEGER INFIX EXPRESSION";
        }
    }
    return result;
}

Object* Evaluator::evalInteger(ASTNode *node, Environment *env) {
    IntegerObject *obj = new IntegerObject();
    obj->value = ((ASTInteger*) node)->value;
    return obj;
}

File: java/Basic2/src/main/java/evaluator/Evaluator.java
package evaluator;

import ast.*;
import environment.Environment;
import objects.B2IntegerObject;
import objects.B2Object;

public class Evaluator {

    public Evaluator() {

    }
    public B2Object eval(ASTNode node, Environment env) {
        switch(node.getType()) {
            case PROGRAM:
                return evalProgram(node, env);
            case EXPRESSION_STATEMENT:
                return evalExpressionStatement(node, env);
            case INFIX_EXPRESSION:
                return evalInfixExpression(node, env);
            case INTEGER:
                return evalIntegerExpression(node, env);
            default:
                System.out.println("UNKNOWM AST NODE EVAL");
                break;
        }
        return null;
    }

    private B2Object evalProgram(ASTNode node, Environment env) {
        ASTProgram program = (ASTProgram) node;
        for (int i = 0; i < program.getStatements().size(); i++) {
            B2Object result = eval(program.getStatements().get(i), env);
            System.out.println(result.inspect());
        }
        return null;
    }

    private B2Object evalExpressionStatement(ASTNode node, Environment env) {
        ASTExpressionStatement stmt = (ASTExpressionStatement) node;
        return eval(stmt.getExpression(), env);
    }

    private B2Object evalInfixExpression(ASTNode node, Environment env) {
        ASTInfixExpression expr = (ASTInfixExpression) node;
        B2IntegerObject left = (B2IntegerObject) eval(expr.getLeft(), env);
        B2IntegerObject right = (B2IntegerObject) eval(expr.getRight(), env);
        B2IntegerObject result = new B2IntegerObject();
        switch(expr.getOperator()) {
            case "+":
                result.setValue(left.getValue() + right.getValue());
                return result;
            case "-":
                result.setValue(left.getValue() - right.getValue());
                return result;
            case "*":
                result.setValue(left.getValue() * right.getValue());
                return result;
            case "/":
                result.setValue(left.getValue() / right.getValue());
                return result;
            default:
                System.out.println("UNKNOWN INFIX EXPRESSION OPERATOR EVAL");
                break;
        }
        return null;
    }

    private B2Object evalIntegerExpression(ASTNode node, Environment env) {
        B2IntegerObject result = new B2IntegerObject();
        ASTInteger integer = (ASTInteger) node;
        result.setValue(integer.getValue());
        return result;
    }
}
